api/envoy/extensions/filters/network/ext_proc/v3/ext_proc.proto

```diff
@@ -0,0 +1,87 @@
+option (xds.annotations.v3.file_status).work_in_progress = true;
+
+// [#protodoc-title: External Processing Filter]
+// [#not-implemented-hide:]
+// External Processing Filter: Process network traffic using an external service.
+// [#extension: envoy.filters.network.ext_proc]
+
+// The Network External Processing filter allows an external service to process raw TCP/UDP traffic
+// in a flexible way using a bidirectional gRPC stream. Unlike the HTTP External Processing filter,
+// this filter operates at the L4 (transport) layer, giving access to raw network traffic.
+//
+// The filter communicates with an external gRPC service that can:
+// * Inspect traffic in both directions
+// * Modify the network traffic
+// * Control connection lifecycle (continue, close, or reset)
+//
+// By using the filter's processing mode, you can selectively choose which data
+// directions to process (read, write or both), allowing for efficient processing.
+message NetworkExternalProcessor {
+  // The gRPC service that will process network traffic.
+  // This service must implement the NetworkExternalProcessor service
+  // defined in the proto file /envoy/service/network_ext_proc/v3/external_processor.proto.
+  config.core.v3.GrpcService grpc_service = 1;
+
+  // By default, if the gRPC stream cannot be established, or if it is closed
+  // prematurely with an error, the filter will fail. Specifically, if the
+  // response headers have not yet been delivered, then it will return a 500
+  // error downstream. If they have been delivered, then instead the HTTP stream to the
+  // downstream client will be reset.
+  // With this parameter set to true, however, then if the gRPC stream is prematurely closed
+  // or could not be opened, processing continues without error.
+  // [#not-implemented-hide:]
+  bool failure_mode_allow = 2;
+
+  // Options for controlling processing behavior.
+  // [#not-implemented-hide:]
+  ProcessingMode processing_mode = 3;
+
+  // Specifies the timeout for each individual message sent on the stream and
+  // when the filter is running in synchronous mode. Whenever
+  // the proxy sends a message on the stream that requires a response, it will
+  // reset this timer, and will stop processing and return an error (subject
+  // to the processing mode) if the timer expires. Default is 200 ms.
+  // [#not-implemented-hide:]
+  google.protobuf.Duration message_timeout = 4 [(validate.rules).duration = {
+    lte {seconds: 3600}
+    gte {}
+  }];
+}
+
+// Options for controlling processing behavior.
+message ProcessingMode {
+  // Defines how traffic should be handled by the external processor
+  enum ProcessingAction {
+    // Send the data to the external processor for processing
+    PROCESS = 0;
+
+    // Skip sending the data to the external processor
+    BYPASS = 1;
+  }
+
+  // Controls whether inbound (read) data from the client is sent to the external processor.
+  // Default: PROCESS
+  ProcessingAction process_read = 1;
```

Comment:
Having separate settings for read and write means that it's possible to configure both of them to BYPASS.  That effectively disables the filter, which doesn't seem very useful.

As an alternative, consider something like this:

```
enum ProcessingMode {
  READ_AND_WRITE = 0;
  READ = 1;
  WRITE = 2;
}
ProcessingMode processing_mode = 1;
```

Reply:
Okay.  Is there any use case where we would want to configure the filter to skip both reads and writes?  If not, maybe we should have the filter reject that config.

Reply:
yes, good point, and I think let filter reject it could be a reasonable approach.

Reply:
Please document in the comments that the filter will reject that config.

Reply:
Done, thanks!